extensions/src/main/scala/ai/h2o/sparkling/extensions/serde/SerializationUtils.scala

```diff
@@ -0,0 +1,172 @@
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+package ai.h2o.sparkling.extensions.serde
+
+import java.sql.Timestamp
+
+import water.AutoBuffer
+
+object SerializationUtils {
+  val EXPECTED_BOOL = 0
+  val EXPECTED_BYTE = 1
+  val EXPECTED_CHAR = 2
+  val EXPECTED_SHORT = 3
+  val EXPECTED_INT = 4
+  val EXPECTED_FLOAT = 5
+  val EXPECTED_LONG = 6
+  val EXPECTED_DOUBLE = 7
+  val EXPECTED_STRING = 8
+  val EXPECTED_TIMESTAMP = 9
+  val EXPECTED_VECTOR = 10
+
+  /**
+    * This is used to inform us that another byte is coming.
+    * That byte can be either {@code MARKER_ORIGINAL_VALUE} or {@code MARKER_NA}. If it's
+    * {@code MARKER_ORIGINAL_VALUE}, that means
+    * the value sent is in the previous data sent, otherwise the value is NA.
+    */
+  private[serde] val NUM_MARKER_NEXT_BYTE_FOLLOWS: Byte = 127
+
+  /**
+    * Same as above, but for Strings. We are using unicode code for CONTROL, which should be very very rare
+    * String to send as usual String data.
+    */
+  private[serde] val STR_MARKER_NEXT_BYTE_FOLLOWS = "\u0080"
+
+  /**
+    * Marker informing us that the data are not NA and are stored in the previous byte
+    */
+  private[serde] val MARKER_ORIGINAL_VALUE = 0
+
+  /**
+    * Marker informing us that the data being sent is NA
+    */
+  private[serde] val MARKER_NA = 1
+
+  def sendIntArray(ab: AutoBuffer, data: Array[Int]): Unit = {
+    ab.putA4(data)
+  }
+
+  def sendDoubleArray(ab: AutoBuffer, data: Array[Double]): Unit = {
+    ab.putA8d(data)
+  }
+
+  def sendBoolean(ab: AutoBuffer, data: Boolean): Unit = {
+    sendBoolean(ab, if (data) 1.toByte
+    else 0.toByte)
+  }
+
+  def sendBoolean(ab: AutoBuffer, boolData: Byte): Unit = {
+    ab.put1(boolData)
+    putMarker(ab, boolData)
+  }
+
+  def sendByte(ab: AutoBuffer, data: Byte): Unit = {
+    ab.put1(data)
+    putMarker(ab, data)
+  }
+
+  def sendChar(ab: AutoBuffer, data: Char): Unit = {
+    ab.put2(data)
+    putMarker(ab, data)
+  }
+
+  def sendShort(ab: AutoBuffer, data: Short): Unit = {
+    ab.put2s(data)
+    putMarker(ab, data)
+  }
+
+  def sendInt(ab: AutoBuffer, data: Int): Unit = {
+    ab.putInt(data)
+    putMarker(ab, data)
+  }
+
+  def sendLong(ab: AutoBuffer, data: Long): Unit = {
+    ab.put8(data)
+    putMarker(ab, data)
+  }
+
+  def sendFloat(ab: AutoBuffer, data: Float): Unit = {
+    ab.put4f(data)
+  }
+
+  def sendDouble(ab: AutoBuffer, data: Double): Unit = {
+    ab.put8d(data)
+  }
+
+  def sendString(ab: AutoBuffer, data: String): Unit = {
+    ab.putStr(data)
+    if (data != null && data == STR_MARKER_NEXT_BYTE_FOLLOWS) ab.put1(MARKER_ORIGINAL_VALUE)
+  }
+
+  def sendTimestamp(ab: AutoBuffer, time: Long): Unit = {
+    sendLong(ab, time)
+  }
+
+  def sendTimestamp(ab: AutoBuffer, data: Timestamp): Unit = {
+    sendLong(ab, data.getTime)
+  }
+
+  private[serde] def sendNA(ab: AutoBuffer, expectedType: Byte): Unit = expectedType match {
```

Comment:
Maybe typo? private[serde]

Reply:
Would you like me to rename `serde` package?

Reply:
Maybe something a little bit more explanatory would help? ( Unless this is a know term which just I don't know, in that case it's ok :D )